tests/periph_spi: add support for printing and testing SPI clock rate…

hugueslarrive opened the pull request:
…s on stm32

<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description
Re-open #14768 because I needed it yesterday...

The current test does not allow to test all baud rates supported by the hardware and driver.
On the nucleo-f334r8 board:
* 0: 100 KHz -> 281250KHz
* 1: 400 KHz -> 281250KHz
* 2: 1 MHz -> 562500KHz
* 3: 5 MHz -> 4.5MHz
* 4: 10 MHz -> 9 MHz

It was not possible to make tests with others supported speeds (1.125 MHz, 2.25 MHz, 18MHz, and 36MHz).

One might think that this is a limitation of the RIOT SPI driver but this is no longer the case for the STM32 CPU family...
<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->


### Testing procedure
<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->


### Issues/PRs references

<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

fjmolinas commented on tests/periph_spi/main.c:
```diff
@@ -99,6 +99,38 @@ static uint32_t _xtimer_diff_usec(xtimer_ticks32_t stop, xtimer_ticks32_t start)
+            printf("CLOCK_APB2 = %luHz\n", (uint32_t)CLOCK_APB2);
+        }
+        else {
+            printf("error: apbbus configuration ");
```
```suggestion
            puts("error: apbbus configuration ");
```

hugueslarrive replied:
Sorry but I gave up this part of the code, anyway thanks for the review.

fjmolinas commented on tests/periph_spi/main.c:
```diff
@@ -99,6 +99,38 @@ static uint32_t _xtimer_diff_usec(xtimer_ticks32_t stop, xtimer_ticks32_t start)
+        printf("\t\tSPI_DEV(%i), ", i);
+        if (spi_config[i].apbbus == APB1) {
+            clk = CLOCK_APB1;
+            printf("CLOCK_APB1 = %luHz\n", (uint32_t)CLOCK_APB1);
```
```suggestion
            printf("CLOCK_APB1 = %"PRIu32"Hz\n", (uint32_t)CLOCK_APB1);
```

hugueslarrive replied:
Sorry but I gave up this part of the code, anyway thanks for the review.

fjmolinas commented on tests/periph_spi/main.c:
```diff
@@ -99,6 +99,38 @@ static uint32_t _xtimer_diff_usec(xtimer_ticks32_t stop, xtimer_ticks32_t start)
+        }
+        else if (spi_config[i].apbbus == APB2) {
+            clk = CLOCK_APB2;
+            printf("CLOCK_APB2 = %luHz\n", (uint32_t)CLOCK_APB2);
```
```suggestion
            printf("CLOCK_APB2 = %"PRIu32"Hz\n", (uint32_t)CLOCK_APB2);
```

hugueslarrive replied:
Sorry but I gave up this part of the code, anyway thanks for the review.

fjmolinas reviewed (commented):
Some inline comments, @benpicco you were reviewing the initial PR, anything missing from that one?

benpicco commented:
My only issue is that if this information is needed, we should have an API for that - not poke CPU configuration in application / test code.

hugueslarrive commented:
> My only issue is that if this information is needed, we should have an API for that - not poke CPU configuration in application / test code.

I totally agree with that.

hugueslarrive commented:
# Some tests:
## nucleo-f746zg
```
init
2021-09-01 04:35:48,096 # init
2021-09-01 04:35:48,101 # usage: init <dev> <mode> <clk> [cs port] [cs pin]
2021-09-01 04:35:48,101 # 	dev:
2021-09-01 04:35:48,103 # 		0: SPI_DEV(0)
2021-09-01 04:35:48,104 # 		1: SPI_DEV(1)
2021-09-01 04:35:48,105 # 	mode:
2021-09-01 04:35:48,109 # 		0: POL:0, PHASE:0 - on first rising edge
2021-09-01 04:35:48,112 # 		1: POL:0, PHASE:1 - on second rising edge
2021-09-01 04:35:48,116 # 		2: POL:1, PHASE:0 - on first falling edge
2021-09-01 04:35:48,120 # 		3: POL:1, PHASE:1 - on second falling edge
2021-09-01 04:35:48,120 # 	clk:
2021-09-01 04:35:48,123 # 		0: SPI_CLK_100KHZ (421875 Hz)
2021-09-01 04:35:48,126 # 		1: SPI_CLK_400KHZ (421875 Hz)
2021-09-01 04:35:48,128 # 		2: SPI_CLK_1MHZ (843750 Hz)
2021-09-01 04:35:48,131 # 		3: SPI_CLK_5MHZ (3375000 Hz)
2021-09-01 04:35:48,133 # 		4: SPI_CLK_10MHZ (6750000 Hz)
2021-09-01 04:35:48,136 # 		n: arbitrary value in Hz
2021-09-01 04:35:48,137 # 	cs port:
2021-09-01 04:35:48,142 # 		Port of the CS pin, set to -1 for hardware chip select
2021-09-01 04:35:48,143 # 	cs pin:
2021-09-01 04:35:48,149 # 		Pin used for chip select. If hardware chip select is enabled,
2021-09-01 04:35:48,153 # 		this value specifies the internal HWCS line
init 0 0 1700000 0 0
2021-09-01 04:36:11,845 # init 0 0 1700000 0 0
2021-09-01 04:36:11,852 # SPI_DEV(0) initialized: mode: 0, clk: 1700000(1687500 Hz),cs_port: 0, cs_pin: 0
> init 0 0 14000000 0 0
2021-09-01 04:37:16,051 # init 0 0 14000000 0 0
2021-09-01 04:37:16,058 # SPI_DEV(0) initialized: mode: 0, clk: 14000000(13500000 Hz),cs_port: 0, cs_pin: 0
> init 0 0 27000000 0 0
2021-09-01 04:37:41,326 # init 0 0 27000000 0 0
2021-09-01 04:37:41,333 # SPI_DEV(0) initialized: mode: 0, clk: 27000000(27000000 Hz),cs_port: 0, cs_pin: 0
> init 0 0 54000000 0 0
2021-09-01 04:37:51,521 # init 0 0 54000000 0 0
2021-09-01 04:37:51,529 # SPI_DEV(0) initialized: mode: 0, clk: 54000000(54000000 Hz),cs_port: 0, cs_pin: 0
> init 0 0 108000000 0 0
2021-09-01 04:38:09,621 # init 0 0 108000000 0 0
2021-09-01 04:38:09,628 # SPI_DEV(0) initialized: mode: 0, clk: 108000000(54000000 Hz),cs_port: 0, cs_pin: 0
```
All clock rates are accessible now.
## esp8266-esp-12x
```
init
2021-09-01 04:41:04,630 # init
2021-09-01 04:41:04,654 # usage: init <dev> <mode> <clk> [cs port] [cs pin]
2021-09-01 04:41:04,654 # 	dev:
2021-09-01 04:41:04,655 # 		0: SPI_DEV(0)
2021-09-01 04:41:04,655 # 	mode:
2021-09-01 04:41:04,656 # 		0: POL:0, PHASE:0 - on first rising edge
2021-09-01 04:41:04,657 # 		1: POL:0, PHASE:1 - on second rising edge
2021-09-01 04:41:04,658 # 		2: POL:1, PHASE:0 - on first falling edge
2021-09-01 04:41:04,659 # 		3: POL:1, PHASE:1 - on second falling edge
2021-09-01 04:41:04,674 # 	clk:
2021-09-01 04:41:04,675 # 		0: SPI_CLK_100KHZ (100000 Hz)
2021-09-01 04:41:04,676 # 		1: SPI_CLK_400KHZ (400000 Hz)
2021-09-01 04:41:04,677 # 		2: SPI_CLK_1MHZ (1000000 Hz)
2021-09-01 04:41:04,678 # 		3: SPI_CLK_5MHZ (5000000 Hz)
2021-09-01 04:41:04,679 # 		4: SPI_CLK_10MHZ (10000000 Hz)
2021-09-01 04:41:04,679 # 	cs port:
2021-09-01 04:41:04,681 # 		Port of the CS pin, set to -1 for hardware chip select
2021-09-01 04:41:04,681 # 	cs pin:
2021-09-01 04:41:04,683 # 		Pin used for chip select. If hardware chip select is enabled,
2021-09-01 04:41:04,685 # 		this value specifies the internal HWCS line
> init 0 0 5 0 0
2021-09-01 04:42:07,981 # init 0 0 5 0 0
2021-09-01 04:42:07,984 # error: invalid bus speed specified
```
The driver do not supports arbitrary bus speed.
## atmega1284p
```
2021-09-01 04:50:41,813 # init
2021-09-01 04:50:41,865 # usage: init <dev> <mode> <clk> [cs port] [cs pin]
2021-09-01 04:50:41,871 # 	dev:
2021-09-01 04:50:41,888 # 		0: SPI_DEV(0)
2021-09-01 04:50:41,895 # 	mode:
2021-09-01 04:50:41,940 # 		0: POL:0, PHASE:0 - on first rising edge
2021-09-01 04:50:41,986 # 		1: POL:0, PHASE:1 - on second rising edge
2021-09-01 04:50:42,031 # 		2: POL:1, PHASE:0 - on first falling edge
2021-09-01 04:50:42,078 # 		3: POL:1, PHASE:1 - on second falling edge
2021-09-01 04:50:42,084 # 	clk:
2021-09-01 04:50:42,117 # 		0: SPI_CLK_100KHZ (62500 Hz)
2021-09-01 04:50:42,150 # 		1: SPI_CLK_400KHZ (250000 Hz)
2021-09-01 04:50:42,181 # 		2: SPI_CLK_1MHZ (500000 Hz)
2021-09-01 04:50:42,213 # 		3: SPI_CLK_5MHZ (2000000 Hz)
2021-09-01 04:50:42,247 # 		4: SPI_CLK_10MHZ (4000000 Hz)
2021-09-01 04:50:42,257 # 	cs port:
2021-09-01 04:50:42,316 # 		Port of the CS pin, set to -1 for hardware chip select
2021-09-01 04:50:42,326 # 	cs pin:
2021-09-01 04:50:42,392 # 		Pin used for chip select. If hardware chip select is enabled,
2021-09-01 04:50:42,440 # 		this value specifies the internal HWCS line
```
The resulting speeds displayed are half of the targeted speeds because the `spi_clk_t enum` from `cpu/atmega_common/periph_cpu.h` has been written assuming a master clock speed of 16 MHz, but this card operates at only 8 MHz. So this which is displayed is what you really get on the SCK PIN.

## Other architectures
I have not the hardware to make tests on other architectures.

maribu commented:
See https://github.com/RIOT-OS/RIOT/pull/15904 as a generic API

MrKevinWeiss commented:
Ping me if you want any [SPI tests](https://github.com/RIOT-OS/RobotFW-tests/tree/master/tests/periph_spi) run on [the HiL CI](https://hil.riot-os.org/jenkins/).  We support clock speed checking, I think there is a bit of a rework that could use this.  ping @MarcelStenzel

maribu commented on boards/frdm-k22f/include/periph_conf.h:
```diff
@@ -141,22 +141,54 @@ static const uart_conf_t uart_config[] = {
+    [9] = {
+        .dev = ADC0,    .pin = GPIO_UNDEF,
+        .chan = 27,     .avg = ADC_AVG_MAX
+    },
```
I agree that this increases readability. Could you split this change out into a separate PR?

maribu commented on cpu/atmega_common/periph/spi.c:
```diff
@@ -29,17 +29,56 @@
+     * right shifting these register value up to 5:
+     * clk = (CLOCK_CORECLOCK / 2) >> (SPR1 << 2 | SPR0 << 1 | ~SPI2X)
+     */
+    uint8_t shift = 0;
+    while (clk << shift < CLOCK_CORECLOCK / 2) {
+        shift++;
+    }
```
```suggestion
    uint8_t shift = 0;
    uint32_t spi_clk = CLOCK_CORECLOCK / 2;
    while (spi_clk > clk) {
        shift++;
        spi_clk >>= 1;
    }
```

IMO this is easier to read, as the MCU will divide down the core clock to generate the SPI clock. Also, this will no longer overflow for high values of `clk`.

maribu commented on cpu/atmega_common/periph/spi.c:
```diff
@@ -78,6 +117,25 @@ void spi_init_pins(spi_t bus)
+spi_clk_t spi_get_clk(spi_t bus, uint32_t freq)
+{
+    (void)bus;
+    /* bound divider from 2 to 128 */
+    if (freq > CLOCK_CORECLOCK / 2) {
+        freq = CLOCK_CORECLOCK / 2;
+    }
```
```suggestion
```

No longer needed if you take the suggestions above.

maribu commented on cpu/atxmega/periph/spi.c:
```diff
@@ -87,6 +130,36 @@ void spi_init_pins(spi_t bus)
     gpio_init(spi_config[bus].mosi_pin, GPIO_OUT);
 }
 
+uint32_t spi_clk_info(spi_t bus, spi_clk_t clk)
+{
+    (void)bus;
+    (void)clk;
+    /* in spi_acquire :
+     * CLK2X    PRESCALER
+     * 1        01          ->  ClkPER / 8
+     * All clock prescalers was set to 1 so ClkPER = CLOCK_CORECLOCK */
+    return CLOCK_CORECLOCK / 8;
+}
+
```
This looks like a leftover to me.

maribu commented on cpu/atxmega/periph/spi.c:
```diff
@@ -87,6 +130,36 @@ void spi_init_pins(spi_t bus)
+    }
+    assert(freq >= CLOCK_CORECLOCK / 128);
+
+    return _clk_shift(freq);
```
```suggestion
    uint8_t shift = _clk_shift(freq);
    return ((~shift & 1) << 7) | (shift >> 1)
```

Maybe prepare the value directly in the format that it can be or'ed into the CTRL register?

maribu reviewed (commented):
The AT(X)mega implementations look good to me. I hope I manage to review the remaining soonish.

maribu commented on drivers/include/periph/spi.h:
```diff
@@ -172,16 +183,18 @@ typedef enum {
-} spi_clk_t;
-#endif
+enum {
+    SPI_CLK_100KHZ  = KHZ(100), /**< drive the SPI bus with 100KHz */
```
Btw: We could implement all backends for now so that `spi_clk_t` will always hold the correct clock divider for the first SPI peripheral and convert it as needed if `spi_acquire()` is called on a bus that is clocked differently.

This would leave some potential for optimization unused, but it would be fully backward compatible. Once the deprecated frequency constants are fully phased out, this optimization potential could be used. Non-breaking API changes are always easier to sell :-)

fjmolinas commented on cpu/cc2538/periph/spi.c:
```diff
@@ -94,19 +97,53 @@ void spi_init_pins(spi_t bus)
     dev(bus)->CR0 = 0;
-    dev(bus)->CPSR = (spi_clk_config[clk].cpsr);
-    dev(bus)->CR0 = ((spi_clk_config[clk].scr << 8) | mode | SSI_CR0_DSS(8));
+    dev(bus)->CPSR = clk >> 8;
```
Why this shift?

fjmolinas commented on cpu/efm32/periph/spi.c:
```diff
@@ -35,6 +35,9 @@
 #include "em_usart.h"
 
+/* DIV_UP is division which rounds up instead of down */
+#define SPI_DIV_UP(a, b)    (((a) + ((b) - 1)) / (b))
```
Could be moved into some `macros/xxx.h` header.

fjmolinas commented on cpu/efm32/periph/spi.c:
```diff
@@ -56,6 +59,69 @@ void spi_init_pins(spi_t bus)
     gpio_init(spi_config[bus].miso_pin, GPIO_IN_PD);
 }
 
+spi_clk_t spi_get_clk(spi_t bus, uint32_t freq)
```
@basilfx could you take a look and test this one?

fjmolinas reviewed (commented):
@MrKevinWeiss could HIL setup help in testing all of this?

fjmolinas commented:
Generally in favor of this one, but its a big change, I don't have an oscilloscope, but I do have a bunch of BOARDs with spi devices attached so I could do testing of that kind. I can do some more testing next tuesday.

MrKevinWeiss commented:
> @MrKevinWeiss could HIL setup help in testing all of this?

This is what is was built for.

@MarcelStenzel was reworking the SPI tests to get around some limitations of PHiLIP and DUT hardware but we can give it a go now.  Pretty soon I can boot up rack 5 and add another 8 boards to the mix.

MrKevinWeiss commented:
[Don't be too alarmed](https://hil.riot-os.org/results/robot_tests/16/overview)

Tests are expected to fail.  There are numbers on the actual SPI clock measurements hidden in the logs.

MrKevinWeiss commented:
[Here](https://hil.riot-os.org/results/robot_tests/16/arduino-due/tests_periph_spi/log#s1-s4-t7-k2-k8)  is an example of logs showing the actual time measurements

fjmolinas commented:
> [Here](https://hil.riot-os.org/results/robot_tests/16/arduino-due/tests_periph_spi/log#s1-s4-t7-k2-k8) is an example of logs showing the actual time measurements

Do you have a link to a master run to compare to?

MrKevinWeiss commented:
[From last night](https://hil.riot-os.org/results/nightly/latest/overview)

MrKevinWeiss commented:
If there is an API change I would have to adapt the robot tests to match that though...

fjmolinas commented:
@hugueslarrive could you rewrite the PR description since this PR is now going further than the initial intent?

fjmolinas commented:
> If there is an API change I would have to adapt the robot tests to match that though...

Yes I think there are new functions here to test, how hard would it be to add new tests for this? @hugueslarrive @maribu could you give @MrKevinWeiss an idea of what kind of new tests would be needed in the HIL? This would make this PR getting in sooooo much easier.

MrKevinWeiss commented:
Ya I can take a look... I was hoping to offload to @MarcelStenzel but I think he doesn't have time to do all my requests.

MrKevinWeiss commented:
fyi the failures are occurring in the nucleos due to the assert in the `spi_get_clk` function.  It seems I cannot do that much exploratory testing of clocks without either overriding or disabling the assert functions.

I am wondering if a min/max would be better than the assert, especially since there are no default speeds per board?

maribu commented:
> I am wondering if a min/max would be better than the assert, especially since there are no default speeds per board?

The API guarantees you that the picked clock will not be higher than the requested to avoid driving the bus out of spec. This is based on the observation that datasheets usually contain an upper bound for the SPI clock, but either no lower bound or a lower bound as low as 100 kHz or so.

So: You should never experience a blown assertion for too high frequencies (the actual clock will just saturate when maxed out). But when you request e.g. 1 Hz as SPI clock, you'll get a blown assertion.

From the user point of view this is IMO ideal and there is no need for min / max.

For testing: I would start with a reasonably high clock (e.g. 1 MHz) and increase by 50%. Only if the returned clock config differs from the last run, it makes sense to actually test. If the clock config hasn't changed for the last 12 iterations (so was increased by more than factor 129), it should be pretty safe to assume that the maximum frequency was reached.

MrKevinWeiss commented:
What about adding an enum or getter function to see the min/max available?

I can think of applications that would try to communicate at a high speed and if failures occur to lower it.  For example, if something needs to communicate through varying EMC conditions.

maribu commented:
OK, you convinced me that adding a getter function for the maximum and minimum SPI frequency makes sense. (It will have to be a function or a function-like macro, since some MCUs support different frequencies for their SPI peripherals. E.g. they might have an high speed QSPI that can be used to connect flash to it that is used as regular SPI in RIOT, and a normal SPI (without Quad-) that runs at a lower clock.)

The reason I was skeptical to this is that I observed that niche features in periph APIs often lead to some of the implementations not implementing it consistently, correctly, or at all. For that reason I'm always in favor of keeping peripheral APIs as slim as possible and rather implement convenience features on top of it.

But I agree that dynamically adapting the clock frequency is something where you don't want to run in an `assert()`, but rather just keep the SPI at the lowest frequency and hope for conditions to improve over time.

I don't think that a getter for a maximum frequency is strictly needed, but as it is trivial to automatically test its correctness, it should be a piece of cake to maintain.

MrKevinWeiss commented:
The assert essentially removed the possibility of setting a lowest frequency.

As long as I have that I am happy.  With this update we should also be able to verify if the frequency now.

I also don't need a max frequency but thought, if we have enums (and it is in fact trivial) then why not.

fjmolinas commented:
@maribu @MrKevinWeiss would be really nice to get this one in this release, is there still some testing missing?

MrKevinWeiss commented:
I still would like the `assert` removed when setting a low frequency.

fjmolinas commented:
@maribu @MrKevinWeiss @hugueslarrive lets not let this change diverge too much, maybe we can directly push the required changes?

- rebase
- remove assert and implement getter functions?
- something else?

maribu commented:
Sounds great. I cannot highlight enough how cool this would be to have in the release :)

gschorcht commented:
At least for EPSs I can say that it works like a charm. I have tested different clocks that work even though ALL SPI tests in [HIL test output](https://hil.riot-os.org/results/robot_tests/177/esp32-wroom-32/tests_periph_spi/console_log.html) failed.

MrKevinWeiss commented:
> Sounds great. I cannot highlight enough how cool this would be to have in the release :)

I agree, this can also allow us to tighten up the tolerances for the SPI clock tests.  Again, the only issue I have is the asserts when trying to set a specific clock rate.

MrKevinWeiss commented:
> even though ALL SPI tests in HIL test output failed

Yup, you should ignore these results.  The default speed for most tests is 100k, since that fails the assertion all tests fail.

gschorcht commented:
It seems that we need testers for different platforms. I would guess that @hugueslarrive does not have the hardware available for all platforms.

MrKevinWeiss commented:
Once the asserts are changed to min/maxing or even return codes, I can rerun and results should be more accurate.

gschorcht commented:
@hugueslarrive ESP implementation has to be changed a bit after the changes for slow APB clock frequencies (2 MHz and 40 MHz) in PR #17413.

gschorcht commented on cpu/esp_common/periph/spi.c:
```diff
@@ -63,6 +63,13 @@
+#define SPI_DIV_UP(a, b)    (((a) + ((b) - 1)) / (b))
+
+/* baud_rate = 80MHz / (spi_clkdiv_pre + 1) / (spi_clkcnt_N + 1) */
+#define SPI_CLK_SRC_FREQ    80000000
```
This frequency depends on APB clock frequencies which is 2 MHz for CPU clock of 2 MHz and 40 MHz for CPU clock of 40 MHz. The following change should work.
```suggestion
#if CONFIG_ESP32_DEFAULT_CPU_FREQ_MHZ == 2
#define SPI_CLK_SRC_FREQ    2000000
#elif CONFIG_ESP32_DEFAULT_CPU_FREQ_MHZ == 40
#define SPI_CLK_SRC_FREQ    40000000
#else
#define SPI_CLK_SRC_FREQ    80000000
#endif
```

OlegHahm commented:
@hugueslarrive, any chance to make the changes for ESP?

MrKevinWeiss commented:
Worst case could someone take it over? It is pretty useful.

MrKevinWeiss commented:
Ping @hugueslarrive are you interested in following up on this?  If not I would like to have a go at it...

hugueslarrive commented:
Let's get back to it

hugueslarrive commented:
Not so easy after 2 years...

> Once the asserts are changed to min/maxing or even return codes, I can rerun and results should be more accurate.

@MrKevinWeiss I tried to replace those asserts with return codes, is that okay?

@gschorcht I can no longer compile tests/periph/spi for esp8266, it fails with:
```
section `.text' will not fit in region `iram1_0_seg'
region `iram1_0_seg' overflowed by 36796 bytes
```
I also have these weird make outputs on some archs (including esp32, esp8266 and msp430):
```
/bin/sh: 1: arithmetic expression: expecting primary: ""
```

I can no longer compile tests/periph/spi for msp430, it fails with:
```
.rodata not found for insert
```

maribu commented:
>I also have these weird make outputs on some archs (including esp32, esp8266 and msp430):
>
>    /bin/sh: 1: arithmetic expression: expecting

please rebase on master, that has been fixed in the meantime

The issue on ESP8266 is that ISRs are kept in RAM (instruction RAM, IRAM), as loaded code from the SPI flash is not feasible there. It seems that to much code is now in the IRAM.

MrKevinWeiss commented on cpu/efm32/periph/spi.c:
```diff
@@ -56,6 +59,69 @@ void spi_init_pins(spi_t bus)
+    clkdiv = clkdiv << 8;
+
+    /* Verify that resulting clock divider is within limits. */
+    EFM_ASSERT(!(clkdiv & ~_USART_CLKDIV_MASK));
```
Hmmm, since a user could theoretically set the frequency, can we avoid the assert and truncate or something?  Since there  can be a difference between the set frequency and the real frequency there would be a verify step anyways if it is important.

hugueslarrive replied:
I simply removed it, as it would never have been triggered anyway, since the limit is already checked at the beginning of the function.

MrKevinWeiss reviewed (commented):
Minor comment, but nice to have some error codes for scanning, thanks.

gschorcht commented on drivers/ds3234/include/ds3234_params.h:
```diff
@@ -47,15 +47,15 @@ extern "C" {
  * @brief   Configure DS3234 devices
  */
-static const ds3234_params_t ds3234_params[] =
+static ds3234_params_t ds3234_params[] =
```
Hm, too bad, this prevents the parameter set from being stored in ROM instead of RAM.

hugueslarrive replied:
For this one I didn't have a device descriptor where to store the spi clock configuration so I fell back on a caching function, I'm not sure that's an improvement...

gschorcht commented on drivers/st7735/include/st7735_params.h:
```diff
@@ -104,7 +107,7 @@ extern "C" {
  * @brief   Configure LCD
  */
-static const lcd_params_t st7735_params[] =
+static lcd_params_t st7735_params[] =
```
Here and in a couple of other drivers. It's a bad idea to remove the `const` keyword in parameter set declarations because it prevents that they are placed in ROM.

The correct approach would be to consider the `spi_freq` in the parameter set as requested frequency and to store the actual frequency in the device descriptor which is then used in `spi_acquire`. This approach would require only 4 bytes of RAM instead of 20 byte as in this driver.

hugueslarrive replied:
Fine, I'll correct it

gschorcht commented on boards/frdm-k22f/include/periph_conf.h:
```diff
@@ -141,22 +143,54 @@ static const uart_conf_t uart_config[] = {
-    [ 5] = { .dev = ADC0, .pin = GPIO_PIN(PORT_B,  1), .chan =  9, .avg = ADC_AVG_MAX }, /* PTB1 (Arduino A1) */
-    [ 6] = { .dev = ADC0, .pin = GPIO_PIN(PORT_C,  1), .chan = 15, .avg = ADC_AVG_MAX }, /* PTC1 (Arduino A2) */
-    [ 7] = { .dev = ADC0, .pin = GPIO_PIN(PORT_C,  2), .chan =  4, .avg = ADC_AVG_MAX }, /* PTC2 (Arduino A3) */
+    [0] = { /* ADC0_DP0 */
```
What's the reason for this unrelated change? 100 characters per line are fine now.

hugueslarrive replied:
100 characters per line on laptop screens are not fine for bad eyes like mine...
It will disappear with the next rebase: https://github.com/RIOT-OS/RIOT/pull/19774

gschorcht commented on cpu/atxmega/periph/spi.c:
```diff
@@ -87,22 +123,46 @@ void spi_init_pins(spi_t bus)
+{
+    (void)bus;
+    /* bound divider to 128 */
+    if(freq < DIV_ROUND_UP(CLOCK_CORECLOCK, 128)) {
```
```suggestion
    if (freq < DIV_ROUND_UP(CLOCK_CORECLOCK, 128)) {
```

gschorcht commented on cpu/atxmega/periph/spi.c:
```diff
@@ -87,22 +123,46 @@ void spi_init_pins(spi_t bus)
+    }
+
+    uint8_t shift = _clk_shift(freq);
+    return (spi_clk_t){
+        .clk = ((~shift & 1) << 7) | (shift >> 1)
+    };
```
Looks a bit wierd.
```suggestion
    return (spi_clk_t){ .clk = ((~shift & 1) << 7) | (shift >> 1) };
```

gschorcht commented on cpu/atxmega/periph/spi.c:
```diff
@@ -87,22 +123,46 @@ void spi_init_pins(spi_t bus)
+int32_t spi_get_freq(spi_t bus, spi_clk_t clk)
+{
+    (void)bus;
+    if (clk.err) { return -EINVAL; }
```
```suggestion
    if (clk.err) {
        return -EINVAL;
    }
```

gschorcht commented on cpu/atxmega/periph/spi.c:
```diff
@@ -87,22 +123,46 @@ void spi_init_pins(spi_t bus)
-    (void)clk;
+
+    assert(bus < SPI_NUMOF);
+    if (clk.err) { return; }
```
```suggestion
    if (clk.err) {
        return -EINVAL;
    }
```

gschorcht commented on cpu/cc2538/periph/spi.c:
```diff
@@ -94,19 +97,58 @@ void spi_init_pins(spi_t bus)
+int32_t spi_get_freq(spi_t bus, spi_clk_t clk)
+{
+    (void)bus;
+    if (clk.err) { return -EINVAL; }
```
```suggestion
    if (clk.err) {
        return -EINVAL;
    }
```

gschorcht commented on cpu/cc2538/periph/spi.c:
```diff
@@ -94,19 +97,58 @@ void spi_init_pins(spi_t bus)
+    (void)cs;
+
     assert((unsigned)bus < SPI_NUMOF);
+    if (clk.err) { return; }
```
```suggestion
    if (clk.err) {
        return -EINVAL;
    }
```

gschorcht commented on cpu/efm32/periph/spi.c:
```diff
@@ -56,6 +59,66 @@ void spi_init_pins(spi_t bus)
+int32_t spi_get_freq(spi_t bus, spi_clk_t clk)
+{
+    (void)bus;
+    if (clk.err) { return -EINVAL; }
```
```suggestion
    if (clk.err) {
        return -EINVAL;
    }
```
Here and at dozends other places.

gschorcht commented on drivers/ili9341/include/ili9341_params.h:
```diff
@@ -95,7 +98,7 @@ extern "C" {
  * @brief   Configure ILI9341
  */
-static const lcd_params_t ili9341_params[] =
+static lcd_params_t ili9341_params[] =
```
Actual SPI frequency should be stored in `lcd_t` instead of making the parameter set non-const.

gschorcht reviewed (changes requested):
Some general comments:

- Instead of making parameter declarations non-const, the SPI frequency in the parameter set should be considered as requested frequency, but the actual SPI frequency should then be stored in the corresponding device descriptor. Removing the `const` keyword from the parameter set declaration has the disadvantage that the parameter set variable (that can be pretty large) is then placed in RAM instead of ROM.

- I am very impressed how many drivers/packages have to be changed because of the impact of this API change, so I wonder if this is really the right approach. Backwards compatibility is essential for user acceptance. Instead of changing all the hundreds of drivers, you should think again about using an approach as already suggested by @maribu in https://github.com/RIOT-OS/RIOT/pull/16727#discussion_r713019732. The PR is already pretty large and still not complete. It is so large that can't almost handle it in my browser. The approach suggested by @maribu would allow to split the PR into the API change and driver changes step by step.
Also the approach mentioned by @maribu in https://github.com/RIOT-OS/RIOT/pull/16727#discussion_r713065620 to cache the last used SPI frequency settings in the backends and to change the settings only if another frequency is used in `spi_acquire` wouldn't be a problem and easy to implement.

- I'm also wondering whether the change of the name from `spi_clk` to `spi_freq` in the parameter sets are really necessary in the first step.

hugueslarrive commented:
@gschorcht Thank you very much for your comments. About the const modifiers, I corrected the drivers from which I had removed it but there are others where I had left it without it causing a compile error so I'll have to see why...

The problem with caching the last used frequency settings in the backend is the case where several devices with different frequencies are used alternatively. Not only this can be a performance killer on architectures with costly prescaler computing, but it's also detrimental to real time, since spi_acquire will take more or less time depending on whether or not another thread has accessed the bus just before. That's why I'm convinced that the caching must be done in each device driver so that spi_acquire remains deterministic.

User acceptance shouldn't be a major issue since this is bus driver is primarily used through other device drivers. It should be transparent in most cases except for those who use their own driver without sharing it.

maribu commented:
Sorry for the merge conflict with MSP430. I started rework all periph drivers for that MCU anyway, e.g. it currently only supports a single SPI bus (and at least the MSP430 F2xx family with the USCI based SPI can support multiple SPI interfaces). (And I also would like to look into sharing the USART/USCI peripheral providing the UART/SPI/I2C in the way it is done in nRF5x; right now the MSP430 x1xx family can use one UART plus either one I2C or one SPI, which frankly is pretty scarce.)

I think this PR should just drop SPI support for the MSP430 and stop worrying about that exotic platform. I will eventually add new SPI, UART and I2C implementations anyway. And `git` makes it pretty easy to base that work on the old code even when it was removed.

benpicco reviewed (commented):
I really don't understand the rationale behind the `.spi_clk` -> `.spi_freq` name change.

On most(?) platforms, `SPI_CLK_5MHZ` is already defined to be `MHZ(5)` today, so this isn't even a semantic change, but creates a lot of busywork.

maribu commented:
> User acceptance shouldn't be a major issue since this is bus driver is primarily used through other device drivers. It should be transparent in most cases except for those who use their own driver without sharing it.

I mostly agree with that. Still, users with drivers used as external modules will feel some pain.

Given that 99% of the use cases will either use a single SPI bus or all SPI buses are compatible with the same prescaler settings, providing the backward compatible constants would turn this into a non-breaking API change except for 1% of the use cases. (It might also be possible to pre-calculate prescaler settings for all SPI buses in a `struct` and pick the right one in `spi_acquire()`. Or to just always return the prescale for one type of SPI periph and convert the prescaler into the prescaler of other types in `spi_acquire()` with something as simple as a bit shift or so. This behavior could then be phased out over time when the backward compatible macros are finally removed.)

In addition to not breaking external modules (for at least 99% of the cases), this would also allow to update the drivers independently as follow ups. This would allow distributing the work (also the review tasks) on more shoulders.

Just to be sure I'm not misunderstood: I'm all in for phasing out `spi_clk_t` constants, especially with the prospect of turning SPI prescaler calculations from run time operations to compile time operations. But I think @gschorcht has a strong point in avoiding breaking changes for users of the SPI bus.

maribu commented on drivers/include/periph/spi.h:
```diff
@@ -127,6 +131,19 @@ typedef uint_fast8_t spi_t;
+ * This structure encapsulates the return code and the SPI clock configuration.
+ */
+#ifndef HAVE_SPI_CLK_T
+typedef struct {
+    int err;        /**< Error code indicating the success or failure of the
+                     *   operation */
+    uword_t clk;    /**< SPI clock configuration */
+} spi_clk_t;
```
Combining an error code with the clock configuration in a `struct` looks pretty non-idiomatic C to me. Rather an `int spi_get_clk(spi_clk_t *dest, spi_t bus, uint32_t freq_hz)` would be more idiomatic to me. But that would conflict with my proposed mode of backward compatibility.

But to be honest: After having seen lots of users of peripheral APIs with return code checking and graceful error handling intended to be done by the caller, I lost all hope that driver developers will ever do this correctly. If `spi_get_clk()` would just `assert()` sane settings, we likely get better error handling for less bytes in `.text`.

Also, the only way to incorrectly call `spi_get_clk()` to my understanding is calling it with a non-existing value for `spi_t bus` or with a frequency that is so low that no prescaler can be found. The former clearly is a bug on the caller side, the latter is also difficult to recover from at runtime: If the device attached to a given hardware SPI bus and that bus just cannot generate a frequency compatible with that device, other than falling back to bit-baning SPI in software I have no idea to handle this at runtime. And I'd rather have the application developer than explicitly use the soft SPI than falling back at runtime.

gschorcht replied:
> Rather an `int spi_get_clk(spi_clk_t *dest, spi_t bus, uint32_t freq_hz)` would be more idiomatic to me.

:+1:

If I am correct, the only error code used in `spi_clk_t::err` is `-EDOM`? And probably no platform will return 0 as `spi_clk_t::clk` from `spi_get_clk`. So why not use as return value 0 in case of error and the clock different from 0 in case of success?

hugueslarrive replied:
The qn908x platform can return 0:
```c
    /* The value stored in DIV is always (divider - 1), meaning that a value of
     * 0 divides by 1. */
    return (spi_clk_t){ .clk = (uint16_t)(divider - 1) };
```
I guess I should not have changed the description from "Opaque type that contains an SPI clock configuration".

The user should not worry about what it contains but just use it to pass the value obtained from `spi_get_clk` to `spi_acquire` or `spi_get_freq` (which returns an int containing the actual frequency value or -EDOM in the idiomatic way).

I will put a better description in the documentation.

gschorcht replied:
> The qn908x platform can return 0:

:thinking: too bad

hugueslarrive commented:
@benpicco 
> I really don't understand the rationale behind the `.spi_clk` -> `.spi_freq` name change.
> 
> On most(?) platforms, `SPI_CLK_5MHZ` is already defined to be `MHZ(5)` today, so this isn't even a semantic change, but creates a lot of busywork.


This is not a name change, spi_freq has been added.


The problem is that spi_acquire needs a prescaler or divider value to be applied in MCU registers, which basically requires the source clock to be divided by the expected frequency, and that doing this computation at each acquire is a big waste of resources. Ideally spi_clk should be renamed spi_div or spi_psc or something like that (which is more related to a period than a frequency) but this won't be the case for all platforms even after this PR so I've left spi_clk for now.


As asking the user to provide a value ready to be applied to the registers is not an option, we provide a getter function that computes this parameter from an expected frequency. We also provide a convenience getter function that returns the actual resulting frequency from this parameter.


So each driver can get the required clock parameter during initialization and store it for later uses.


With this approach you can share the bus between several devices with different frequencies and access them in turn without any overhead.


This was not a marginal issue, as great efforts were made to mitigate it in some major implementations such as stm32.

benpicco commented:
> The problem arises when several different devices use the same bus, i.e. a storage device and a network device.

but in that case all real-time assumptions are out of the window already as any of the two devices might block the bus for an arbitrary amount of time.

maribu commented:
> > The problem arises when several different devices use the same bus, i.e. a storage device and a network device.
> 
> but in that case all real-time assumptions are out of the window already as any of the two devices might block the bus for an arbitrary amount of time.

Not really. Real time capable doesn't mean fast, just "in time". Reducing the overhead of `spi_acquire()` allows splitting larger SPI transactions into multiple smaller ones without too much overhead. This in turn would reduce the worst case time a higher prio thread has to wait for a lower prio thread to release the bus.

IMO reducing the overhead of `spi_acquire ()` is very much beneficial for real time apps.

hugueslarrive commented on cpu/esp32/periph/spi.c:
```diff
@@ -263,11 +263,66 @@ void spi_deinit_pins(spi_t bus)
+    if (freq > apb_clk / 5) {
+        LOG_TAG_ERROR("spi", "APB clock rate (%"PRIu32" Hz) has to be at "
+                      "least 5 times SPI clock rate (%"PRIu32" Hz)\n",
+                      apb_clk, freq);
```
@gschorcht I moved this code as is from spi_acquire but I wonder why APB clock rate has to be at least 5 times SPI clock rate. From the reference manual section 7.4: _"The maximum output clock frequency of ESP32 GP-SPI master is f apb /2"_ and even _"When the SPI_CLK_EQU_SYSCLK bit in register SPI_CLOCK_REG is set to 1, and the other bits are set to 0, SPI output clock frequency is f apb."_

gschorcht replied:
This is what I had to learn when I was implementing the support for core clock frequencies lower than 80 MHz down to 2 MHz. For example with
```
CFLAGS='-DCONFIG_ESP32_DEFAULT_CPU_FREQ_MHZ=40' BOARD=esp32-wroom-32 make -j8 -C tests/periph/spi flash
```
and wired MOSI-MISO I get for 10 MHz:
```
> init 0 0 4 0 5
> send hallo
Sent bytes
   0    1    2    3    4 
  0x68 0x65 0x6c 0x6c 0x6f
    h    e    l    l    o 

Received bytes
   0    1    2    3    4 
  0x34 0x32 0xb6 0x36 0x37
    4    2   ??    6    7 
```
while it works for 5 MHz.

I have the same problem for 2 MHz core clock frequency. While it works for 400 kHz SPI clock frequency, it doesn't work for 1 MHz SPI clock frequency.

hugueslarrive replied:
Note that 0x3432b63637 = 0x68656c6c6f >> 1. I observed the same issue on esp8266 for the highest frequency (80MHz with spi_clk_equ_sysclk).

It seems to be linked to spi_miso_delay_mode which should be set to 0 for frequencies higher than clkapb/4 according to the manual section 7.4.2.
